openquake/hazardlib/shakemap/conditioned_gmfs.py

```diff
@@ -0,0 +1,903 @@
+            for i in range(len(conditioning_imts)):
+                T_D[i * nss: (i + 1) * nss, i + 1] = tau_D[
+                    i * nss: (i + 1) * nss, 0]
+        # The raw residuals
+        zeta_D = yD - mu_yD
+
+        # Compute the station data within-event covariance matrix
+        rho_WD_WD = compute_spatial_cross_correlation_matrix(
+            station_sitecol_filtered,
+            station_sitecol_filtered,
+            conditioning_imts,
+            conditioning_imts,
+            spatial_correl,
+            cross_correl_within,
+        )
+        phi_D_flat = phi_D.flatten()
+        cov_WD_WD = numpy.linalg.multi_dot(
+            [numpy.diag(phi_D_flat), rho_WD_WD, numpy.diag(phi_D_flat)]
+        )
+        # Add on the additional variance of the residuals
+        # for the cases where the station data is uncertain
+        numpy.fill_diagonal(cov_WD_WD, numpy.diag(cov_WD_WD) + var_addon_D)
+
+        # Get the (pseudo)-inverse of the station data within-event covariance
+        # matrix
+        cov_WD_WD_inv = numpy.linalg.pinv(cov_WD_WD)
+
+        # # The normalized between-event residual and its variance (for the
+        # # observation points)
+        # # Engler et al. (2022) equations 12 and 13; assumes between event
+        # # residuals are perfectly cross-correlated
+        # var_H_y2 = 1.0 / (
+        #     1.0 + numpy.linalg.multi_dot([tau_y2.T, cov_W2_W2_inv, tau_y2])
+        # )
+        # mu_H_y2 = numpy.linalg.multi_dot(
+        #   [tau_y2.T, cov_W2_W2_inv, zeta]) * var_H_y2
+        # The more generic equations B8 and B9 from Appendix B are used instead
+        # requiring the computation of the covariance matrix Σ_HD_HD, which is
+        # just the matrix of cross-correlations for the observed IMTs, since
+        # H is the normalized between-event residual
+        corr_HD_HD = cross_correl_between._get_correlation_matrix(
+            bracketed_imts)
+
+        # Using Bayes rule, compute the posterior distribution of the
+        # normalized between-event residual H|YD=yD, employing
+        # Engler et al. (2022), eqns B8 and B9 (also B18 and B19),
+        # H|Y2=y2 is normally distributed with mean and covariance:
+        cov_HD_HD_yD = numpy.linalg.pinv(
+            numpy.linalg.multi_dot([T_D.T, cov_WD_WD_inv, T_D])
+            + numpy.linalg.pinv(corr_HD_HD)
+        )
+        mu_HD_yD = numpy.linalg.multi_dot(
+            [cov_HD_HD_yD, T_D.T, cov_WD_WD_inv, zeta_D])
+
+        # Compute the distribution of the conditional between-event
+        # residual B|Y2=y2
+        mu_BD_yD = T_D @ mu_HD_yD
+        cov_BD_BD_yD = numpy.sqrt(
+            numpy.diag(numpy.linalg.multi_dot([T_D, cov_HD_HD_yD, T_D.T])))
+
+        # Get the nominal bias and its variance as the means of the
+        # conditional between-event residual mean and covariance
+        nominal_bias_mean = numpy.mean(mu_BD_yD)
+        # ASK: why are we computing nominal_bias_stddev if it is not used??
```

Comment:
@raoanirudh , why we are ignoring nominal_bias_mean?

Reply:
Now also logging the nominal_bias_stddev